### ISSUE TITLE
An instance of TimeWithZone class does NOT quote `TO_DATE' function on Rails 4.2.x

### ISSUE THREAD

**Issue by koic** [rsim/oracle-enhanced#755]
Is this intentionally changed?

## rails 4.1.13 / activerecord-oracle_enhanced-adapter (1.5.6) (Expected)

Call ActiveRecord::Base.where method with an instance of TimeWithZone argument.

```ruby
> created_at = Model.first.created_at
> Model.where(created_at: created_at)
  Model Load (5.1ms)  SELECT "MODELS".* FROM "MODELS"  WHERE "MODELS"."CREATED_AT" = TO_DATE('2015-12-28 03:55:04','YYYY-MM-DD HH24:M
I:SS')
=> [#<Model id: ***snip***>]
```

Returning an Array of matched instances, as I had expected.

## rails 4.2.5 / activerecord-oracle_enhanced-adapter (1.6.5) (Actual)

Ditto. Call ActiveRecord::Base.where method with an instance of TimeWithZone argument.

```ruby
> created_at = Model.first.created_at
> Model.where(created_at: created_at)
  Model Load (11.8ms)  SELECT "MODELS".* FROM "MODELS" WHERE "MODELS"."CREATED_AT" = :a1  [["created_at", 2015-12-28 03:55:04 UTC]]
=> []
```

This SQL does NOT quoted `TO_DATE` function. Then returning an empty Array. (ummm)

If that is NOT intentional, then patch is difficult to me...

**Comment by yahonda** [comment]
Thanks for reporting. Would you create a minimum and entire test case to reproduce your issue.

```ruby
# Activate the gem you are reporting the issue against.
gem 'activerecord', '4.2.5'
gem 'activerecord-oracle_enhanced-adapter', '1.6.5'
require 'active_record'
require 'minitest/autorun'
require 'logger'
require 'active_record/connection_adapters/oracle_enhanced_adapter'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# Set Oracle enhanced adapter specific connection parameters
DATABASE_NAME = ENV['DATABASE_NAME'] || 'orcl'
DATABASE_HOST = ENV['DATABASE_HOST']
DATABASE_PORT = ENV['DATABASE_PORT']
DATABASE_USER = ENV['DATABASE_USER'] || 'oracle_enhanced'
DATABASE_PASSWORD = ENV['DATABASE_PASSWORD'] || 'oracle_enhanced'
DATABASE_SYS_PASSWORD = ENV['DATABASE_SYS_PASSWORD'] || 'admin'

CONNECTION_PARAMS = {
  :adapter => "oracle_enhanced",
  :database => DATABASE_NAME,
  :host => DATABASE_HOST,
  :port => DATABASE_PORT,
  :username => DATABASE_USER,
  :password => DATABASE_PASSWORD
}

ActiveRecord::Base.establish_connection(CONNECTION_PARAMS)

ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, :force => true do |t|
  end

  create_table :comments, :force => true do |t|
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_association_stuff
    post = Post.create!
    post.comments << Comment.create!

    assert_equal 1, post.comments.count
    assert_equal 1, Comment.count
    assert_equal post.id, Comment.first.post.id
  end
end
```

Thanks.

**Comment by koic** [comment]
Thank you for your quick reply.

I wrote a test case for the issue.

```ruby
# Activate the gem you are reporting the issue against.
gem 'activerecord', '4.2.5'
gem 'activerecord-oracle_enhanced-adapter', '1.6.5'
require 'active_record'
require 'minitest/autorun'
require 'logger'
require 'active_record/connection_adapters/oracle_enhanced_adapter'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# Set Oracle enhanced adapter specific connection parameters
DATABASE_NAME = ENV['DATABASE_NAME'] || 'orcl'
DATABASE_HOST = ENV['DATABASE_HOST']
DATABASE_PORT = ENV['DATABASE_PORT']
DATABASE_USER = ENV['DATABASE_USER'] || 'oracle_enhanced'
DATABASE_PASSWORD = ENV['DATABASE_PASSWORD'] || 'oracle_enhanced'
DATABASE_SYS_PASSWORD = ENV['DATABASE_SYS_PASSWORD'] || 'admin'

CONNECTION_PARAMS = {
  :adapter => "oracle_enhanced",
  :database => DATABASE_NAME,
  :host => DATABASE_HOST,
  :port => DATABASE_PORT,
  :username => DATABASE_USER,
  :password => DATABASE_PASSWORD
}

ActiveRecord::Base.establish_connection(CONNECTION_PARAMS)

ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, :force => true do |t|
    t.timestamps null: false
  end
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_find_by_created_at
    post = Post.create!

    created_at = post.created_at

    assert_equal post, Post.find_by(created_at: created_at)
  end
end
```

This test case will fail.

Just for your information, test passes on the following versions.

```ruby
gem 'activerecord', '4.1.13'
gem 'activerecord-oracle_enhanced-adapter', '1.5.6'
```

Please contact me if you find anything unclear in the above. Thanks.

**Comment by yahonda** [comment]
Thanks it reproduces.

**Comment by yahonda** [comment]
## Which version to reproduce
* It reproduces with CRuby 2.3.0 and JRuby 9.0.4.0 both.
* It reproduces with the latest Rails 4-2-stable branch and Oracle enhanced adapter 1.6.5.

## Rails version differences


* Rails 4.1 `created_at` value is casted by to_date function, no bind value used.

```
Post Load (21.5ms)  SELECT  "POSTS".* FROM "POSTS"  WHERE "POSTS"."CREATED_AT" = TO_DATE('2016-01-08 14:33:47','YYYY-MM-DD HH24:MI:SS') AND ROWNUM <= 1
```

* Rails 4.2 `created_at` value is handled as bind value.

```
Post Load (4.7ms)  SELECT  "POSTS".* FROM "POSTS" WHERE "POSTS"."CREATED_AT" = :a1 AND ROWNUM <= 1   [["created_at", 2016-01-08 14:28:05 UTC]]
```

I think it is an expected behavior to use bind values, the problem could be data type differences, of course this is my assumption it could be wrong. I'll update this assumption later.

**Comment by yahonda** [comment]
I think it is an expected behavior to use bind values, the problem could be data type differences, of course this is my assumption it could be wrong. I'll update this assumption later.

My assumption is this issue happens when the bind value datatype is `TIMESTAMP WITH TIME ZONE` for the column `DATE`, which does not have time zone information. It also does not have fractional seconds, but I do not think it causes this problem since Rails do not pass fractional seconds to `created_at` value.


* `created_at` is `Time` class object

```ruby
(byebug) created_at.class
Time
```

* `CREATED_AT` column is defined as `DATE` sql type, Oracle `DATE` type does not have time zone information and fractional seconds.

```sql

SQL> desc posts
 Name                                      Null?    Type
 ----------------------------------------- -------- ----------------------------
 ID                                        NOT NULL NUMBER(38)
 CREATED_AT                                NOT NULL DATE
 UPDATED_AT                                NOT NULL DATE

SQL>

```

* When bind_param called, this `Time` class value is bound as `OCI8::BindType::Time`

```ruby
[140, 149] in /home/yahonda/git/oracle-enhanced/lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb
   140:             when :decimal
   141:               @raw_cursor.bind_param(position, BigDecimal.new(value.to_s))
   142:             else
   143:               require 'byebug'
   144:               byebug
=> 145:               @raw_cursor.bind_param(position, value)
   146:             end
   147:           end
   148:         end
   149:
(byebug) @raw_cursor.bind_param(position, value)
#<OCI8::Cursor:0x005612814a8d70 @bind_handles={1=>#<OCI8::BindType::Time:0x0056127f75acf0>}, @define_handles=[], @column_metadata=[], @names=nil, @con=#<OCI8:ORACLE_ENHANCED>, @max_array_size=nil>
(byebug) puts position
1
nil
(byebug) puts value
2016-01-08 14:40:47 UTC
nil
(byebug) puts value.class
Time
nil
(byebug)
```
* sql_trace using 10046 level 12, which shows this bind value data type is `oacdty=181`, which should be `TIMESTAMP WITH TIME ZONE`

```sql
 Bind#0
  oacdty=181 mxl=13(13) mxlc=00 mal=00 scl=09 pre=00
  oacflg=01 fl2=9000000 frm=00 csi=00 siz=16 off=0
  kxsbbbfp=7f335aae5ff0  bln=13  avl=13  flg=05
  value=08-JAN-16 04.47.59.810019072 PM +00:00
```

**Comment by yahonda** [comment]
I have been trying to find how to workarounds and/or fixes like this but unable to find any yet. 

* Change `CREATED_AT` and `UPDATED_AT` data types to timezone aware datatype, such as `TIMESTAMP WITH TIME ZONE` 

* Cast `created_at` value to remove time zone information

**Comment by yahonda** [comment]
(Removed my previous updates since I might have misunderstood and could give some wrong directions for this issue)

**Comment by koic** [comment]
I understand. I need to arrange information.
Please wait a moment.

**Comment by yahonda** [comment]
@koic It was my bad. I think your report includes necessary information. My assumption was that this issue is related with bind value data type difference between `TIMESTAMP WITH TIME ZONE` and `DATE`. but I found that this issue still reproduces without using bind values by setting `  :prepared_statements => false`.

It creates the sql `Post Load (5.5ms)  SELECT  "POSTS".* FROM "POSTS" WHERE "POSTS"."CREATED_AT" = TO_DATE('2016-01-11 01:50:28','YYYY-MM-DD HH24:MI:SS') AND ROWNUM <= 1   [["created_at", 2016-01-11 01:50:28 UTC]]`

```ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  #gem 'activerecord', '4.2.5'
  gem 'activerecord', github: 'yahonda/rails', branch: '4-2-stable'
  #gem 'activerecord-oracle_enhanced-adapter', '1.6.5'
  gem 'activerecord-oracle_enhanced-adapter', github: 'rsim/oracle-enhanced', branch: 'master'
  gem 'ruby-oci8'
  gem 'minitest'
  gem 'byebug'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'
require 'active_record/connection_adapters/oracle_enhanced_adapter'
require 'byebug'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# Set Oracle enhanced adapter specific connection parameters
DATABASE_NAME = ENV['DATABASE_NAME'] || 'orcl'
DATABASE_HOST = ENV['DATABASE_HOST']
DATABASE_PORT = ENV['DATABASE_PORT']
DATABASE_USER = ENV['DATABASE_USER'] || 'oracle_enhanced'
DATABASE_PASSWORD = ENV['DATABASE_PASSWORD'] || 'oracle_enhanced'
DATABASE_SYS_PASSWORD = ENV['DATABASE_SYS_PASSWORD'] || 'admin'

CONNECTION_PARAMS = {
  :adapter => "oracle_enhanced",
  :database => DATABASE_NAME,
  :host => DATABASE_HOST,
  :port => DATABASE_PORT,
  :username => DATABASE_USER,
  :password => DATABASE_PASSWORD,
  :prepared_statements => false
}

#ActiveRecord::Base.default_timezone = :local
ActiveRecord::Base.default_timezone = :utc
ActiveRecord::Base.establish_connection(CONNECTION_PARAMS)

ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, :force => true do |t|
    t.timestamps null: false
  end
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_find_by_created_at
    post = Post.create!

    created_at = post.created_at

    assert_equal created_at, Post.find_by!(created_at: created_at)
  end
end
```

```ruby
D, [2016-01-11T01:50:29.634755 #9328] DEBUG -- :   SQL (13.1ms)  INSERT INTO "POSTS" ("CREATED_AT", "UPDATED_AT", "ID") VALUES (TO_DATE('2016-01-11 01:50:28','YYYY-MM-DD HH24:MI:SS'), TO_DATE('2016-01-11 01:50:28','YYYY-MM-DD HH24:MI:SS'), 10000)  [["created_at", 2016-01-11 01:50:28 UTC], ["updated_at", 2016-01-11 01:50:28 UTC], ["id", 10000]]
D, [2016-01-11T01:50:29.646488 #9328] DEBUG -- :   Post Load (5.5ms)  SELECT  "POSTS".* FROM "POSTS" WHERE "POSTS"."CREATED_AT" = TO_DATE('2016-01-11 01:50:28','YYYY-MM-DD HH24:MI:SS') AND ROWNUM <= 1   [["created_at", 2016-01-11 01:50:28 UTC]]
F

Finished in 4.154250s, 0.2407 runs/s, 0.2407 assertions/s.

  1) Failure:
BugTest#test_find_by_created_at [rep42.rb:67]:
--- expected
+++ actual
@@ -1 +1 @@
-2016-01-11 01:50:28 UTC
+#<Post id: 10000, created_at: "2016-01-11 01:50:28", updated_at: "2016-01-11 01:50:28">


1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```

**Comment by yahonda** [comment]
Please ignore my previous update. I did my bad twice. There is a workaround by setting `  :prepared_statements => false` in your database.yml not to execute prepared statements.



```ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  #gem 'activerecord', '4.2.5'
  gem 'activerecord', github: 'yahonda/rails', branch: '4-2-stable'
  #gem 'activerecord-oracle_enhanced-adapter', '1.6.5'
  gem 'activerecord-oracle_enhanced-adapter', github: 'rsim/oracle-enhanced', branch: 'master'
  gem 'ruby-oci8'
  gem 'minitest'
  gem 'byebug'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'
require 'active_record/connection_adapters/oracle_enhanced_adapter'
require 'byebug'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# Set Oracle enhanced adapter specific connection parameters
DATABASE_NAME = ENV['DATABASE_NAME'] || 'orcl'
DATABASE_HOST = ENV['DATABASE_HOST']
DATABASE_PORT = ENV['DATABASE_PORT']
DATABASE_USER = ENV['DATABASE_USER'] || 'oracle_enhanced'
DATABASE_PASSWORD = ENV['DATABASE_PASSWORD'] || 'oracle_enhanced'
DATABASE_SYS_PASSWORD = ENV['DATABASE_SYS_PASSWORD'] || 'admin'

CONNECTION_PARAMS = {
  :adapter => "oracle_enhanced",
  :database => DATABASE_NAME,
  :host => DATABASE_HOST,
  :port => DATABASE_PORT,
  :username => DATABASE_USER,
  :password => DATABASE_PASSWORD,
  :prepared_statements => false
}

#ActiveRecord::Base.default_timezone = :local
ActiveRecord::Base.default_timezone = :utc
ActiveRecord::Base.establish_connection(CONNECTION_PARAMS)

ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, :force => true do |t|
    t.timestamps null: false
  end
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_find_by_created_at
    post = Post.create!

    created_at = post.created_at

    assert_equal post, Post.find_by!(created_at: created_at)
  end
end
```

**Comment by yahonda** [comment]
Disabling prepared statements at database.yml may cause some performance problem. Although Oracle enhanced adapter sets `cursort_sharing = force` by default which can relax the performance about parsing sql statement cpu time. 

I think (finally) think this issue is related with bind value data type difference between `TIMESTAMP WITH TIME ZONE` and `DATE`. Will keep investigating.

**Comment by koic** [comment]
Thank you for your explanation.
That really helps. I understood very well.

**Comment by yahonda** [comment]
It would take some more time to investigate, then I'm updating the current summary of my assumption.

My assumption is this issue happens when the bind value datatype is `TIMESTAMP WITH TIME ZONE` for the column `DATE`, which does not have time zone information. `DATE` also does not have fractional seconds, but I do not think it causes this problem since Rails do not pass fractional seconds to `created_at` value.


* `created_at` is `Time` class object

```ruby
(byebug) created_at.class
Time
```

* `CREATED_AT` column is defined as `DATE` sql type, Oracle `DATE` type does not have time zone information and fractional seconds.

```sql

SQL> desc posts
 Name                                      Null?    Type
 ----------------------------------------- -------- ----------------------------
 ID                                        NOT NULL NUMBER(38)
 CREATED_AT                                NOT NULL DATE
 UPDATED_AT                                NOT NULL DATE

SQL>

```

* When bind_param called, this `Time` class value is bound as `OCI8::BindType::Time`

```ruby
[140, 149] in /home/yahonda/git/oracle-enhanced/lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb
   140:             when :decimal
   141:               @raw_cursor.bind_param(position, BigDecimal.new(value.to_s))
   142:             else
   143:               require 'byebug'
   144:               byebug
=> 145:               @raw_cursor.bind_param(position, value)
   146:             end
   147:           end
   148:         end
   149:
(byebug) @raw_cursor.bind_param(position, value)
#<OCI8::Cursor:0x005612814a8d70 @bind_handles={1=>#<OCI8::BindType::Time:0x0056127f75acf0>}, @define_handles=[], @column_metadata=[], @names=nil, @con=#<OCI8:ORACLE_ENHANCED>, @max_array_size=nil>
(byebug) puts position
1
nil
(byebug) puts value
2016-01-08 14:40:47 UTC
nil
(byebug) puts value.class
Time
nil
(byebug)
```
* sql_trace using 10046 level 12, which shows this bind value data type is `oacdty=181`, which should be `TIMESTAMP WITH TIME ZONE`

```sql
 Bind#0
  oacdty=181 mxl=13(13) mxlc=00 mal=00 scl=09 pre=00
  oacflg=01 fl2=9000000 frm=00 csi=00 siz=16 off=0
  kxsbbbfp=7f335aae5ff0  bln=13  avl=13  flg=05
  value=08-JAN-16 04.47.59.810019072 PM +00:00
```

Then from Oracle point of view, this prepared statement sql statement uses bind values whose data type is `TIMESTAMP WITH TIME ZONE` for the `CREATED_AT` column whose data type is `DATE`.

If we can change the bind value type to `DATE` or change `CREATED_AT` data type to `TIMESTAMP WITH TIME ZONE` this issue can be addressed.

**Comment by yahonda** [comment]
@kubo Would you do me a favor? I'd like to know about ruby-oci8 bind value data type and column data type since I'm wondering if my assumption that is binding `TIMESTAMP WITH TIME ZONE` for `DATE` column may be related with this issue.

**Comment by yahonda** [comment]
I have found a quick and dirty fix to address this. At least the testcase works fine.
It removes timezone information from the value when bind value data type is datetime.

https://github.com/yahonda/oracle-enhanced/commit/b26f1b6c00cdc0e5aca915ac7c43d5433bf994e6

Notes:
- No JDBC connection fixed yet
- datetime / time data type differences not considered in detail

Ref: http://stackoverflow.com/questions/9132337/how-can-i-remove-the-zone-from-a-datetime-value

**Comment by yahonda** [comment]
I think this issue will be resolved by Oracle enhance adapter 1.7 since it does not reproduce using Oracle enhanced adapter current master branch. Mainly because `timestamps` are mapped to `timestamp` Oracle data type in Rails 5. 

Would you test using Oracle enhanced adapter 1.7.0.beta to reproduce this issue with your application. After upgrading from Rails 4.2 to Rails 5, I think you need to modify sql datatype from `date` to `timestamp` or using Rails 5 attribute API to map `created_at` and `updated_at` as :datetime.

**Comment by koic** [comment]
Thanks for your imformation. I will confirm it, so please wait a moment.

**Comment by yahonda** [comment]
Based on discussion at RubyKaigi, I've got to know it still reproduces when `TZ` environment value is set.  Would you see if it still reproduces with this branch.

```ruby
gem 'activerecord-oracle_enhanced-adapter', github: 'yahonda/oracle-enhanced', branch: 'diag755_tz'
```

Thanks.

**Comment by koic** [comment]
This issue does not reproduce with `diag755_tz` branch, it was resolved.

Many thanks.

**Comment by yahonda** [comment]
Thanks for testing. I'm going to add some specs for this change and open a pull request to master branch.

**Comment by yahonda** [comment]
Closing this issue since Oracle enhanced adapter 1.7.3 has been released including #998. It addresses this issue for Rails 5.0 and Oracle enhanced adapter 1.7. Unfortunately simply backporting #998 into release16 branch will not fix it for Rails 4.2 and Oracle enhanced adapter 1.6.

For Rails 4.2 would you keep using [oracle-enhanced-monky_patch_755/](https://github.com/koic/oracle-enhanced-monky_patch_755/) . Of course,  if you find some way to open a pull request into release16 branch I'm happy to accept it .

Thanks for reporting it with the reproduce test case.

**Comment by koic** [comment]
Noted with thanks 🙏
